// bao-wecom/src/common/component.js
import basicBehavior from "../mixins/basic";
function mapKeys(source, target, map) {
    Object.keys(map).forEach(function (key) {
        if (source[key]) {
            target[map[key]] = source[key];
        }
    });
}
export default function BcomComponent(BcomOptions) {
    var options = {};
    mapKeys(BcomOptions, options, {
        data: 'data',
        props: 'properties',
        watch: 'observers',
        mixins: 'behaviors',
        methods: 'methods',
        beforeCreate: 'created',
        created: 'attached',
        mounted: 'ready',
        destroyed: 'detached',
        classes: 'externalClasses',
        lifetimes: "lifetimes"
    });
    // add default externalClasses
    options.externalClasses = options.externalClasses || [];
    options.externalClasses.push('custom-class');
    // add default behaviors
    options.behaviors = options.behaviors || [];
    options.behaviors.push(basicBehavior);
    // add relations
    var relation = BcomOptions.relation;
    if (relation) {
        options.relations = relation.relations;
        options.behaviors.push(relation.mixin);
    }
    // map field to form-field behavior
    if (BcomOptions.field) {
        options.behaviors.push('wx://form-field');
    }
    // add default options
    if (BcomOptions.options) {
        options.options = Object.assign({
            multipleSlots: true,
            addGlobalClass: true,
        }, BcomOptions.options);
    } else {
        options.options = {
            multipleSlots: true,
            addGlobalClass: true,
        };
    }
    Component(options);
}
